tests/unit/metrics/test_ffwd.py

```diff
@@ -0,0 +1,208 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2018 Spotify AB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numbers
+
+import pytest
+
+from gordon.metrics import ffwd
+
+
+def test_ffwd_protocol_connection_made(mocker):
+    """Message sends and connection closes."""
+    mock_transport = mocker.Mock()
+    message = 'Testing message!'
+
+    protocol = ffwd.FfwdClientProtocol(message)
+    protocol.connection_made(mock_transport)
+
+    mock_transport.sendto.assert_called_once_with(message)
+    mock_transport.close.assert_called_once_with()
+
+
+@pytest.mark.asyncio
+async def test_udp_client_send(mocker):
+    """Sends serialized metric using FfwdClientProtocol."""
+    mock_loop = mocker.Mock()
+
+    actual_func = None
+    actual_addr = None
+
+    async def mock_create_datagram_endpoint(func, remote_addr):
+        nonlocal actual_func
+        nonlocal actual_addr
+        actual_func = func
+        actual_addr = remote_addr
+
+    mock_loop.create_datagram_endpoint = mock_create_datagram_endpoint
+
+    udp_client = ffwd.UDPClient('1.2.3.4', '5678', mock_loop)
+    await udp_client.send({'an': 'object'})
+
+    actual_protocol = actual_func()
+    assert isinstance(actual_protocol, ffwd.FfwdClientProtocol)
+    assert ('1.2.3.4', '5678') == actual_addr
+    assert b'{"an": "object"}' == actual_protocol.message
+
+
+@pytest.fixture
+def mock_udp_client_and_metrics_sent(mocker):
+    metrics_sent = []
+
+    async def mock_send(metric):
+        nonlocal metrics_sent
+        metrics_sent.append(metric)
+
+    mock_udp_client = mocker.Mock()
+    mock_udp_client.send = mock_send
+    return mock_udp_client, metrics_sent
+
+
+@pytest.fixture
+def timer_and_metrics_sent(mock_udp_client_and_metrics_sent):
+    mock_udp_client, metrics_sent = mock_udp_client_and_metrics_sent
+    metric = {'a': 'metric', 'value': None}
+    timer = ffwd.FfwdTimer(metric, mock_udp_client)
+    return timer, metrics_sent
+
+
+@pytest.mark.asyncio
+async def test_ffwdtimer_manual(timer_and_metrics_sent):
+    """Times when operated manually."""
+    timer, metrics_sent = timer_and_metrics_sent
+    await timer.start()
+    assert isinstance(timer._start_time, numbers.Number)
+
+    await timer.stop()
+    assert len(metrics_sent) == 1
+    assert 'metric' == metrics_sent[0]['a']
+    assert 0 < metrics_sent[0]['value']
```

Comment:
I feel like this and the same line below would be a lot easier to read as foo > 0.  But if you're going to do it this way, the length comparison should be in the other order.

Reply:
I agree it would be easier to read (which is probably why I messed up the len() one, since I usually do it that way first out of habit), but our [contributing guidelines](https://github.com/spotify/gordon/blob/develop/CONTRIBUTING.rst#code) say we should do it the other way around. Will fix!

Reply:
So just to clarify though...  Does 0 count as the 'expected' value if it's a comparison?

Reply:
Yes, because that's what we expect the value to be. Even though the guidelines only show the `expected == actual` case, the condition is part of the expectation. We expect `metrics_sent[0]['a']` *to be equal to* `'metric'`, just like we expect `0` *to be less than* `metrics_sent[0]['value']`  (of course it reads better out loud in English as you initially suggested with (actual, expected_condition) rather than (expected_condition, actual)

Reply:
@ocsw @descent3me This isn't part of the feedback about this PR: It looks like there's some confusion around the `expected == actual` convention and I admit that I'm confused by it sometimes as well. Should we consider changing this guideline?

I think my confusion stems from the "_of course it reads better out loud in English_" point plus from the fact that all conditionals follow the "out loud in English" pattern (`if x > 0:` and not `if 0 < x:`).

Reply:
The logic I've heard for this boils down, IIRC, to: usually the expected value is just a variable, whereas the actual is usually a (possibly long) call, so this way you can find the expected without hunting.  I do wonder if there are other reasons; I've been told this before but it definitely can feel weird, especially with comparisons.  I don't remember which order I used for those.

Reply:
https://stackoverflow.com/questions/9331259/unit-testing-why-is-the-expected-argument-always-first-in-equality-tests

Reply:
One of the goals of setting the standard and enforcing it is when your tests fail, it's clear that you know what is on each side of the equality. If we aren't consistent all the way through, I find it extremely confusing when hunting down bugs that I thought were in my code, when actually it's a typo in the expected.

Whether or not we thing expected == actual or actual == expected reads better, we choose one way and stick to it.